packages/@aws-cdk/example-construct-library/lib/example-resource.ts

```diff
@@ -0,0 +1,501 @@
+   * surface a method that allows you to conditionally
+   * add a statement to that Role if it's known.
+   * This is just a convenience,
+   * so that clients of your interface don't have to check {@link role} for null.
+   * Many such methods in the CDK return void;
+   * you can also return a boolean indicating whether the permissions were in fact added
+   * (so, when {@link role} is not null).
+   */
+  addToRolePolicy(policyStatement: iam.PolicyStatement): boolean;
+
+  /**
+   * An example of a method that grants the given IAM identity
+   * permissions to this resource
+   * (in this case - read permissions).
+   */
+  grantRead(identity: iam.IGrantable): iam.Grant;
+
+  /**
+   * Add a CloudWatch rule that will use this resource as the source of events.
+   * Resources that emit events have a bunch of methods like these,
+   * that allow different resources to be triggered on various events happening to this resource
+   * (like item added, item updated, item deleted, ect.) -
+   * exactly which methods you need depends on the resource you're modeling.
+   */
+  onEvent(id: string, options?: events.OnEventOptions): events.Rule;
+
+  /**
+   * Standard method that allows you to capture metrics emitted by this resource,
+   * and use them in dashboards and alarms.
+   * The details of which metric methods you should have of course depends on the
+   * resource that is being modeled.
+   */
+  metricSth(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+}
+
+/**
+ * A common abstract superclass that implements the {@link IExampleResource} interface.
+ * We often have these classes to share code between the {@link ExampleResource}
+ * class and the {@link IExampleResource} instances returned from methods like
+ * {@link ExampleResource.fromExampleResourceName}.
+ * It has to extend the Resource class from the core module.
+ *
+ * Notice that the class is not exported - it's not part of the public API of this module!
+ */
+abstract class ExampleResourceBase extends core.Resource implements IExampleResource {
+  // these stay abstract at this level
+  public abstract readonly exampleResourceArn: string;
+  public abstract readonly exampleResourceName: string;
+  public abstract readonly role?: iam.IRole;
+  // this property is needed for the iam.IGrantable interface
+  public abstract readonly grantPrincipal: iam.IPrincipal;
+  // This is needed for the ec2.IConnectable interface
+  // allow subclasses to write this field.
+  // JSII requires all member starting with an underscore to be annotated with '@internal'.
+  /** @internal */
+  protected _connections: ec2.Connections | undefined;
+
+  /** Implement the ec2.IConnectable interface, using the _connections field. */
+  public get connections(): ec2.Connections {
+    if (!this._connections) {
+      throw new Error('An imported ExampleResource cannot manage its security groups');
+    }
+    return this._connections;
```

Comment:
I wonder if the right pattern here shouldn't be to implement `ec2.IConnectable` on `ExampleResource`  only (not on `IExampleResource`) so you get a type error instead of a runtime one.

Reply:
The thing is, you can have a `fromExampleResourceAttributes` that allows you specify the `securityGroups`, and then even an imported `IExampleResource` can implement `IConnectable`. Example: [`ecs.Cluster`](https://github.com/aws/aws-cdk/blob/7ed60b8a5d42e93e556e3b6e9ee3618931747ac2/packages/%40aws-cdk/aws-ecs/lib/cluster.ts#L59-L59) (yes, I know `ICluster` technically doesn't implement `IConnectable`, but effectively it does, as it defines a `readonly connections: ec2.Connections;` property).